go-fuzz/internal/writerset/writerset.go

```diff
@@ -0,0 +1,98 @@
+// Package writerset implements a mechanism to add and remove writers from a construct
+// similar to io.MultiWriter. Vendored from github.com/stephens2424/writerset
+package writerset
+
+import (
+	"io"
+	"net/http"
+	"sync"
+)
+
+// WriterSet wraps multiple writers like io.MultiWriter, but such that individual
+// writers are easy to add or remove as necessary.
+type WriterSet struct {
+	m  map[io.Writer]chan error
+	mu *sync.Mutex
+}
+
+// New initializes a new empty writer set.
+func New() *WriterSet {
+	return &WriterSet{
+		m:  make(map[io.Writer]chan error),
+		mu: &sync.Mutex{},
+	}
+}
+
+// Add ensures w is in the set.
+func (ws *WriterSet) Add(w io.Writer) <-chan error {
+	ws.mu.Lock()
+	defer ws.mu.Unlock()
+
+	c, ok := ws.m[w]
+	if ok {
+		return c
+	}
+
+	c = make(chan error, 1)
+	ws.m[w] = c
+	return c
+}
+
+// Contains determines if w is in the set.
+func (ws *WriterSet) Contains(w io.Writer) bool {
+	ws.mu.Lock()
+	defer ws.mu.Unlock()
+
+	_, ok := ws.m[w]
+	return ok
+}
+
+// Remove ensures w is not in the set.
+func (ws *WriterSet) Remove(w io.Writer) {
+	ws.mu.Lock()
+	defer ws.mu.Unlock()
+	delete(ws.m, w)
+}
+
+// Write writes data to each underlying writer. If an error occurs on an underlying writer,
+// that writer is removed from the set. The error will be sent on the channel created when
+// the writer was added.
+func (ws *WriterSet) Write(b []byte) (int, error) {
+	var err error
+
+	sumBytes := 0
+	n := 0
+
+	ws.mu.Lock()
+	defer ws.mu.Unlock()
+
+	for w, c := range ws.m {
+		bs, err := w.Write(b)
+		if err != nil {
+			c <- err
+			close(c)
+			delete(ws.m, w)
+		}
+		sumBytes += bs
+		n++
+	}
+
+	if n == 0 {
+		return 0, err
```

Comment:
This and the following return statement do not look correct to me.

Caller should not be affected by presence/absence of registered writers.
However, when writers are present you return some non-zero amount of bytes; while when there are no writers you return 0. What should caller do in that case? Retry write? If there are no registered writers, we still have successfully written len(b) bytes to zero underlying writers.
Also that average number of bytes looks weird. I don't see how it can be ever useful.
Just return (len(b), nil) unconditionally.